src/librustc/session/config.rs

```diff
@@ -2304,11 +2341,37 @@ pub fn build_session_options_and_crate_config(
             );
         };
 
+
+        // Externsl crates start out public,
+        // and become private if we later see
+        // an '--extern-private' key. They never
+        // go back to being public once we've seen
+        // '--extern-private', so we logical-AND
+        // their current and new 'public' value together
```

Comment:
So, according to this logic the `public` property is attached to an entry in the `Externs` map rather than to each individual location.
This means `Externs` should look something like this
```rust
pub struct Externs(BTreeMap<String, ExternEntry>);

#[derive(Clone, Hash, Eq, PartialEq, Ord, PartialOrd, Debug)]
pub struct ExternEntry {
    pub locations: BTreeSet<Option<String>>,
    pub is_private_dep: bool,
}
```

Reply:
I changed `public` to `is_private_dep` to avoid the same thing being represented differently in different structures and requiring negation.

Reply:
See https://github.com/rust-lang/rust/pull/59335#discussion_r271322246 - I believe the current layout is correct.